File: examples/python/part-8/25-security-by-design-checklist.py
# Illustrative examples for notes/part-8-practice/25-security-by-design-checklist.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any


class Severity(str, Enum):
    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"


class Status(str, Enum):
    YES = "Yes"
    PARTIAL = "Partial"
    NO = "No"
    NA = "N/A"


@dataclass
class Item:
    id: str
    title: str
    severity: Severity
    status: Status
    evidence: list[str] = field(default_factory=list)
    owner: str = ""
    review_date: datetime = field(default_factory=lambda: datetime.min)
    reason: str = ""


def validate(items: list[Item]) -> None:
    for item in items:
        if not item.id or not item.title:
            raise ValueError("checklist item has required empty fields")

        if not item.severity or not item.status:
            raise ValueError(f"checklist item has no severity or status: {item.id}")

        if item.severity == Severity.HIGH and item.status == Status.YES and not item.evidence:
            raise ValueError(f"high severity item marked Yes without evidence: {item.id}")

        if item.severity == Severity.HIGH and item.status == Status.NO:
            raise ValueError(f"high severity item is No and blocks release: {item.id}")

        if item.status == Status.PARTIAL and not item.owner:
            raise ValueError(f"partial item has no owner: {item.id}")

        if item.status == Status.NA and not item.reason:
            raise ValueError(f"N/A item requires reason: {item.id}")


def export(items: list[Item]) -> str:
    validate(items)
    payload: list[dict[str, Any]] = []
    for item in items:
        row = asdict(item)
        row["severity"] = item.severity.value
        row["status"] = item.status.value
        if item.review_date != datetime.min:
            row["review_date"] = item.review_date.isoformat()
        else:
            row.pop("review_date", None)
        payload.append(row)
    return json.dumps(payload, indent=2)

File: examples/typescript/part-8/25-security-by-design-checklist.ts
// Illustrative examples for notes/part-8-practice/25-security-by-design-checklist.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum Severity {
  High = "High",
  Medium = "Medium",
  Low = "Low",
}

enum Status {
  Yes = "Yes",
  Partial = "Partial",
  No = "No",
  NA = "N/A",
}

interface Item {
  id: string;
  title: string;
  severity: Severity;
  status: Status;
  evidence?: string[];
  owner?: string;
  reviewDate?: Date;
  reason?: string;
}

function validate(items: Item[]): void {
  for (const item of items) {
    if (!item.id || !item.title) {
      throw new Error("checklist item has required empty fields");
    }

    if (!item.severity || !item.status) {
      throw new Error(`checklist item has no severity or status: ${item.id}`);
    }

    if (item.severity === Severity.High && item.status === Status.Yes && (!item.evidence || item.evidence.length === 0)) {
      throw new Error(`high severity item marked Yes without evidence: ${item.id}`);
    }

    if (item.severity === Severity.High && item.status === Status.No) {
      throw new Error(`high severity item is No and blocks release: ${item.id}`);
    }

    if (item.status === Status.Partial && !item.owner) {
      throw new Error(`partial item has no owner: ${item.id}`);
    }

    if (item.status === Status.NA && !item.reason) {
      throw new Error(`N/A item requires reason: ${item.id}`);
    }
  }
}

function exportItems(items: Item[]): string {
  validate(items);
  return JSON.stringify(items, null, 2);
}
